### PR TITLE
Fix updated Euler angle computation for crystal plasticity models

### LINKED ISSUE #19473: Crystal rotation and Euler angles update
## Reason

- The crystal rotation should be computed the elastic part of the deformation gradient
- We need a convenient way to output Euler angles of crystals during deformation

## Design

- Fix the calculation of crystal rotation in the ComputeMultipleCrystalPlasticityStress class
- Add a new material class to compute the updated Euler angles during the simulation

## Impact

 
Improve crystal plasticity modeling capability.

### PR DESCRIPTION
**Pull request by dewenyushu**
Closes #19473

### CODE REVIEW

File: modules/tensor_mechanics/doc/content/source/materials/crystal_plasticity/ComputeUpdatedEulerAngle.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+!media tensor_mechanics/crystal_plasticity/Inverse_Pole_Fig.png
+    id=IPF_Orients
+    caption=Inverse pole figure of a singe crystal undergone uniaxial tension. Three cases with different initial orientations are demonstrated. The initial orientations are $[111]$, $[112]$, and $[123]$ from the left to the right.
+    style=display:block;margin-left:auto;margin-right:auto;width:90%
+
+The inverse pole figures are plotted using [MTEX](https://github.com/mtex-toolbox/mtex) under matlab. One can reproduce the data that correspond to the above results by running [this test](modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_euler_angle.i) with a larger end_time (50 seconds).
+
+The MOOSE developer would like to acknowledge Dr. Alankar Alankar and Dr. Ritam Chatterjee's contributions in this verification case.
```
Comment (jiangwen84) [inline modules/tensor_mechanics/doc/content/source/materials/crystal_plasticity/ComputeUpdatedEulerAngle.md:29]:
```suggestion
The MOOSE developer would like to acknowledge Dr. Alankar Alankar and Ritam Chatterjee's contributions to this verification case.
```

File: modules/tensor_mechanics/doc/content/source/materials/crystal_plasticity/ComputeUpdatedEulerAngle.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+!media tensor_mechanics/crystal_plasticity/Inverse_Pole_Fig.png
+    id=IPF_Orients
+    caption=Inverse pole figure of a singe crystal undergone uniaxial tension. Three cases with different initial orientations are demonstrated. The initial orientations are $[111]$, $[112]$, and $[123]$ from the left to the right.
+    style=display:block;margin-left:auto;margin-right:auto;width:90%
+
+The inverse pole figures are plotted using [MTEX](https://github.com/mtex-toolbox/mtex) under matlab. One can reproduce the data that correspond to the above results by running [this test](modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_euler_angle.i) with a larger end_time (50 seconds).
```
Comment (jiangwen84) [inline modules/tensor_mechanics/doc/content/source/materials/crystal_plasticity/ComputeUpdatedEulerAngle.md:27]:
```suggestion
The inverse pole figures are plotted using MATLAB tool box [MTEX](https://github.com/mtex-toolbox/mtex). The above results can be reproduced  by running [this test](modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_euler_angle.i) with  end_time = 50 (seconds).
```

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeUpdatedEulerAngle.C [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+void
+ComputeUpdatedEulerAngle::computeQpProperties()
+{
+  computeEulerAngleFromRotationMatrix(_updated_rotation[_qp], _updated_euler_angle[_qp]);
+}
+
+void
+ComputeUpdatedEulerAngle::computeEulerAngleFromRotationMatrix(const RankTwoTensor & rot,
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeUpdatedEulerAngle.C:47]: Did you take a look at `EulerAngles` class? I just found that class. I am not sure if that is useful for you or not. We might want to reduce any code duplications if possible.

**Review by jiangwen84 — CHANGES_REQUESTED** [review]
Looks good. I left a few minor suggestions.

There are several classes that calculate/use Euler angles, e.g. `EulerAngles`, `ComputeElasticityTensorCP`. It would be great if we can unify those codes.

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeUpdatedEulerAngle.C (hunk 3 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeUpdatedEulerAngle.C:47]: Thanks Wen! 

I remember taking a look into the `EulerAngles` class some time ago. Looks like it aims at transforming the Quaternion notation to Euler angle, not transforming the rotation matrix to Euler angle. 

Eigen might be able to transform rotation matrix to Quaternion then to Euler angle though. Let me take a closer look and see if we can simplify this part of the code

**Comment by dewenyushu** [comment]
> Looks good. I left a few minor suggestions.
> 
> There are several classes that calculate/use Euler angles, e.g. `EulerAngles`, `ComputeElasticityTensorCP`. It would be great if we can unify those codes.

Thanks Wen! It is a good idea to unify the Euler angles calculation in other classes. I'll need to look closer into this. Let me create a new issue

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeUpdatedEulerAngle.C (hunk 3 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeUpdatedEulerAngle.C:47]: I tried to make use of the existing `EuerAngles` class. The `Eigen` library enables simpler transformation from rotation matrix to Euler angles, so I do not have to implement the transformation separately, which is good.

One thing is that I'll need to transform the rotation matrix material property (declared as RankTwoTensor in the CP base classes) to Eigen::Matrix (see below), which does not seem ideal. I do not have a good idea to address this now without changing things in the CP base classes. It maybe better to improve this while addressing #19497 if that's okay.

**Comment by dewenyushu** [comment]
@jiangwen84 when you have the time, let me know what you think about the change. Really appreciate it!

File: modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeUpdatedEulerAngle.C [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                                                              RealVectorValue & euler_angle)
+{
+  // transform RankTwoTensor to Eigen::Matrix
+  Eigen::Matrix<Real, 3, 3> rot_mat;
+
+  for (unsigned int i = 0; i < 3; ++i)
+    for (unsigned int j = 0; j < 3; ++j)
+      rot_mat(i, j) = rot(i, j);
+
+  EulerAngles ea(rot_mat);
```
Comment (jiangwen84) [inline modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeUpdatedEulerAngle.C:57]: Without adding new `initialize` function and constructor  `EulerAngles(const Eigen::Matrix<Real, 3, 3> & rotation);` , would the following codes work?
```
Eigen::Quaternion<Real> q(rot_mat);
EulerAngles ea(q);
```

Comment (dewenyushu) [reply modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeUpdatedEulerAngle.C:57]: Thanks Wen! It surely would work. Let me revert some of the changes.

My intention was to make it more useful by having the transformation from rotation matrix to EulerAngles available in the `EulerAngles` class. But we can figure it out later if this is necessary.

File: modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_euler_angle.i [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    component = 2
+    execute_on = timestep_end
+  []
+[]
+
+[BCs]
+  [Periodic]
+    [all]
+      variable = 'disp_x'
+      auto_direction = 'z'
```
Comment (jiangwen84) [inline modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_euler_angle.i:60]: Could you remind me why we only need `disp_x` here? Not `disp_y`?

Comment (dewenyushu) [reply modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_euler_angle.i:60]: I went over the discussion earlier with Ritam. He experimented with different combinations of variables, only the `disp_x` gave the correct inverse pole figure. I simply kept what he had in the periodic BC in this test. 

Having both `disp_x` and `disp_y` changes the Euler angles values a bit, but the pole figures looks fairly similar to what I have in the documentation (generated with `disp_x` only).

Comment (jiangwen84) [reply modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_euler_angle.i:60]: Thanks for your explanation.

**Review by jiangwen84 — APPROVED** [review]

**Pull request merged**